File: src/net/zomis/tttultimate/games/TTClassicControllerWithGravity.java
package net.zomis.tttultimate.games;

import net.zomis.tttultimate.TTBase;

public class TTClassicControllerWithGravity extends TTClassicController {

	public TTClassicControllerWithGravity(TTBase board) {
		super(board);
	}
	
	@Override
	public boolean isAllowedPlay(TTBase tile) {
		boolean sup = super.isAllowedPlay(tile);
		if (!sup)
			return false;
		
		TTBase parent = tile.getParent();
		TTBase below = parent.getSub(tile.getX(), tile.getY() + 1);
		if (below == null)
			return true;
		return below.isWon();
	}

}

File: uttt-common/src/main/kotlin/net/zomis/tttultimate/games/TTClassicControllerWithGravity.kt
package net.zomis.tttultimate.games

import net.zomis.tttultimate.TTBase

class TTClassicControllerWithGravity(board: TTBase) : TTClassicController(board) {

    override fun isAllowedPlay(tile: TTBase): Boolean {
        val sup = super.isAllowedPlay(tile)
        if (!sup)
            return false

        val parent = tile.parent
        val below = parent!!.getSub(tile.x, tile.y + 1) ?: return true
        return below.isWon
    }

}
